File: 97.Interleaving-String.py
# https://leetcode.com/problems/interleaving-string/description/
#
# algorithms
# Hard (26.32%)
# Total Accepted:    95.3k
# Total Submissions: 362k
# beats 100.0% of python submissions
# DFS，精髓是用一个is_visited保存曾经遍历过的可能，避免无效循环


class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        len_s1, len_s2, len_s3 = len(s1), len(s2), len(s3)

        if len_s3 != len_s1 + len_s2:
            return False

        is_visited = {}

        def resursive(i, j, k):
            if i == len_s1:
                return s2[j:] == s3[k:]
            if j == len_s2:
                return s1[i:] == s3[k:]
            if (i, j) in is_visited:
                return is_visited[(i, j)]

            res = False

            if s1[i] == s3[k]:
                res |= resursive(i + 1, j, k + 1)
            if s2[j] == s3[k]:
                res |= resursive(i, j + 1, k + 1)

            is_visited[(i, j)] = res

            return res

        return resursive(0, 0, 0)

File: 97.Interleaving-String.java
import java.util.HashMap;

// https://leetcode.com/problems/interleaving-string/
//
// algorithms
// Hard (27.87%)
// Total Accepted: 110,126
// Total Submissions: 395,156
// beats 92.99% of java submissions


class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int len1 = s1.length(), len2 = s2.length(), len3 = s3.length();

        if (len1 + len2 != len3) {
            return false;
        }

        HashMap<String, Boolean> map = new HashMap<>();

        return recursive(s1, s2, s3, 0, 0, 0, map);
    }

    public boolean recursive(String s1, String s2, String s3, int idx1, int idx2, int idx3,
            HashMap<String, Boolean> map) {
        int len1 = s1.length(), len2 = s2.length(), len3 = s3.length();

        if (idx3 == len3) {
            return true;
        }

        String key = String.valueOf(idx1) + ":" + String.valueOf(idx2);
        if (map.containsKey(key)) {
            return map.get(key);
        }

        boolean res = false;

        if (idx1 == len1) {
            res = s2.substring(idx2).equals(s3.substring(idx3));
        } else if (idx2 == len2) {
            res = s1.substring(idx1).equals(s3.substring(idx3));
        } else {
            if (s1.charAt(idx1) == s3.charAt(idx3)) {
                res = recursive(s1, s2, s3, idx1 + 1, idx2, idx3 + 1, map);
            }
            if (!res && s2.charAt(idx2) == s3.charAt(idx3)) {
                res = recursive(s1, s2, s3, idx1, idx2 + 1, idx3 + 1, map);
            }
        }

        map.put(key, res);

        return res;
    }
}